File: c++/7 - String/Manacher.cpp
Devuelve un vector p donde, para cada i, p[i] es igual al largo del palindromo mas largo con centro en i.
Tener en cuenta que el string debe tener el siguiente formato: %#s[0]#s[1]#...#s[n-1]#$ (s es el string original y n es el largo del string)

vector<int> manacher(string s) {
    int n = s.size();
    vector<int> p(n, 0);
    int c = 0, r = 0;
    for (int i = 1; i < n-1; i++) {
        int j = c - (i-c) ;
        if (r > i) p[i] = min(r-i , p[j]);
        while (s[i+1+p[i]] == s[i-1-p[i]]) 
            p[i]++;
        if (i+p[i] > r) {
            c = i;
            r = i+p[i];
        }
    }
    return p;
}

File: java/7 - String/Manacher.java
ArrayList<Integer> manacher(String S) {
	int n = S.length();
	ArrayList<Integer> P = new ArrayList(n);
	int C = 0, R = 0;
	for (int i = 1; i < n-1; i++) {
		int j = C - (i - C) ;
		if (R > i) P.set(i, Math.min(R - i , P.get(j)));
		while (S.charAt(i + 1 + P.get(i)) == S.charAt(i - 1 - P.get(i)))
                    P.set(i, P.get(i)+1);
		if (i + P.get(i) > R) {
			C = i;
			R = i + P.get(i);
		}
	}
	return P;
}
